src/llmq/quorums_dkgsession.cpp

```diff
@@ -0,0 +1,1328 @@
+                logger.Batch("%s complained about us", member->dmn->proTxHash.ToString());
+            }
+        }
+        if (!m->complaints.empty()) {
+            receivedCount++;
+        }
+    }
+
+    logger.Batch("received and relayed complaint. received=%d", receivedCount);
+}
+
+void CDKGSession::VerifyAndJustify(CDKGPendingMessages& pendingMessages)
+{
+    if (!AreWeMember()) {
+        return;
+    }
+
+    CDKGLogger logger(*this, __func__);
+
+    std::set<uint256> justifyFor;
+
+    for (const auto& m : members) {
+        if (m->bad) {
+            continue;
+        }
+        if (m->badMemberVotes.size() >= (size_t)params.dkgBadVotesThreshold) {
+            logger.Batch("%s marked as bad as %d other members voted for this", m->dmn->proTxHash.ToString(), m->badMemberVotes.size());
+            MarkBadMember(m->idx);
+            continue;
+        }
+        if (m->complaints.empty()) {
+            continue;
+        }
+        if (m->complaints.size() != 1) {
+            logger.Batch("%s sent multiple complaints", m->dmn->proTxHash.ToString());
+            MarkBadMember(m->idx);
+            continue;
+        }
+
+        auto& qc = complaints.at(*m->complaints.begin());
+        if (qc.complainForMembers[myIdx]) {
+            justifyFor.emplace(qc.proTxHash);
+        }
+    }
+
+    logger.Flush();
+    if (!justifyFor.empty()) {
+        SendJustification(pendingMessages, justifyFor);
+    }
+}
+
+void CDKGSession::SendJustification(CDKGPendingMessages& pendingMessages, const std::set<uint256>& forMembers)
+{
+    CDKGLogger logger(*this, __func__);
+
+    assert(AreWeMember());
+
+    logger.Batch("sending justification for %d members", forMembers.size());
+
+    CDKGJustification qj;
+    qj.llmqType = (uint8_t)params.type;
+    qj.quorumHash = pindexQuorum->GetBlockHash();
+    qj.proTxHash = myProTxHash;
+    qj.contributions.reserve(forMembers.size());
+
+    for (size_t i = 0; i < members.size(); i++) {
+        auto& m = members[i];
+        if (!forMembers.count(m->dmn->proTxHash)) {
+            continue;
+        }
+        logger.Batch("justifying for %s", m->dmn->proTxHash.ToString());
+
+        CBLSSecretKey skContribution = skContributions[i];
+
+        if (i != myIdx && ShouldSimulateError("justify-lie")) {
+            logger.Batch("lying for %s", m->dmn->proTxHash.ToString());
+            skContribution.MakeNewKey();
+        }
+
+        qj.contributions.emplace_back(i, skContribution);
+    }
+
+    if (ShouldSimulateError("justify-omit")) {
+        logger.Batch("omitting");
+        return;
+    }
+
+    qj.sig = activeMasternodeManager->OperatorKey()->Sign(qj.GetSignHash());
+
+    logger.Flush();
+
+    quorumDKGDebugManager->UpdateLocalSessionStatus(params.type, [&](CDKGDebugSessionStatus& status) {
+        status.sentJustification = true;
+        return true;
+    });
+
+    pendingMessages.PushPendingMessage(-1, qj, MSG_QUORUM_JUSTIFICATION);
```

Comment:
The `sentJustification=true` (same as the other status updates) could not be true if `PushPendingMessage` return early due a surpassed messages limit.

Might be good to by-pass the limit if the message was created locally (nodeId == -1).

-- For a follow-up PR, this one is already full --

Reply:
A node shouldn't be allowed to send more messages, as that would be bad behavior (see `ReceiveMessage`/`PreVerifyMessage`).
In that case, anyway, `sentWhatever` would already be true.